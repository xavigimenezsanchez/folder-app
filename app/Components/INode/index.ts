import INode, { IINode } from "./INode";

export default INode;
export { INode };
export type { IINode };
