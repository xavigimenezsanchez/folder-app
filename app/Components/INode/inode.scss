.inode {
  position: relative;
  top: 1em;
  -ms-user-select: none;
  -webkit-user-select: none;
  user-select: none;
  &__child {
    display: flex;
    margin: 0.5em;
  }
}
