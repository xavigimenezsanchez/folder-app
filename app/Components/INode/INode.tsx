import { useState, useEffect, MouseEvent } from "react";
import uniqueId from "lodash.uniqueid";
import { Icon, IconNames } from "../Icons";
import { isImage, isText } from "../../utils";
import * as Axios from "axios";
import "./inode.scss";

const axios = Axios.default;
interface IINode {
  path: string;
  contextMenuStatus: boolean;
}

const INode = ({ path, contextMenuStatus }: IINode) => {
  const [children, setChildren] = useState([]);
  useEffect(() => {
    axios.get(`/api/dir?path=${path}`).then((request) => {
      setChildren(
        request.data.map((node) => {
          return { ...node, id: uniqueId(path) };
        })
      );
    });
  }, []);
  const toggleFolder = (index: number, event: MouseEvent<HTMLInputElement>) => {
    event.stopPropagation();
    if (!contextMenuStatus) {
      // If context menu is open a folder must be not toggled
      if (children[index].isDirectory) {
        const childrenCopy = [...children];
        childrenCopy[index].open = !childrenCopy[index].open;
        setChildren(childrenCopy);
      }
    }
  };
  const getClassName = (isDirectory: boolean): string =>
    "inode__child " + (isDirectory ? "directory" : "file");

  const getIconName = (name: string): IconNames =>
    isImage(name)
      ? IconNames.image
      : isText(name)
      ? IconNames.text
      : IconNames.file;

  const getArrowIconName = (open: boolean): IconNames =>
    open ? IconNames.arrowDown : IconNames.arrowRight;

  return (
    <div className="inode">
      {children.map((node, index) => (
        <div
          key={`inode-${node.id}`}
          inode-path={path + (!node.isDirectory ? node.name : "")}
          className={getClassName(node.isDirectory)}
          onClick={(event: MouseEvent<HTMLInputElement>) =>
            toggleFolder(index, event)
          }
        >
          {node.isDirectory && <Icon iconName={getArrowIconName(node.open)} />}
          {!node.isDirectory && <Icon iconName={getIconName(node.name)} />}
          <div key={`inode-name-${node.id}`}>{node.name}</div>
          {node.open && (
            <INode
              path={`${path}${node.name}/`}
              contextMenuStatus={contextMenuStatus}
            />
          )}
        </div>
      ))}
    </div>
  );
};

export default INode;
export type { IINode };
