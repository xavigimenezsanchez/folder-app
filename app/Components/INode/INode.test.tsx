import React from "react";
import renderer from "react-test-renderer";
import INode from "./INode";

describe("INode", () => {
  const component = renderer.create(<INode path="/"></INode>);
  console.log(component.toJSON());
  test("test", () => {
    let iNode = component.toJSON();
    expect(iNode).toMatchSnapshot();
  });
});
