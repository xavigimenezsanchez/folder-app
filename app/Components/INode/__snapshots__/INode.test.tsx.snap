// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`INode test 1`] = `
<div>
  Inode
</div>
`;
